File: Gombang/[LeetCode75] - [DFS]/1372_Longest_ZigZag_Path_in_a_Binary_Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    const int LEFT = -1;
    const int RIGHT = 1;

    int answer = 0;

    void dfs(TreeNode* root, int count, int direction)
    {
        if (root == nullptr)
            return;

        // 지금까지 저장된 가장 큰 answer값과 새로 들어온 count값의 크기 비교.
        answer = max(answer, count + 1);

        // 현재 노드로 들어오는 탐색 방향이 LEFT였다면,
        if (direction == LEFT)
        {
            dfs(root->right, count + 1, RIGHT);
            dfs(root->left, 0, LEFT); // 현재 노드에서부터 dfs를 다시 시작.
        }
        else
        {
            dfs(root->left, count + 1, LEFT);
            dfs(root->right, 0, RIGHT); // 현재 노드에서부터 dfs를 다시 시작.
        }
    }

    int longestZigZag(TreeNode* root) {

        dfs(root->left, 0, LEFT);
        dfs(root->right, 0, RIGHT);

        return answer;
    }
};

File: Seori/LeetCode 75/Binary Tree_DFS/1372. Longest ZigZag Path in a Binary Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def longestZigZag(self, root: Optional[TreeNode]) -> int:
        def dfs(root: Optional[TreeNode], direction: str, depth: int) -> int:
            # 재귀 종료조건
            if root is None:
                return depth

            # dfs 탐색을 하면서, 그 지점이 지그재그의 시작점일 수 있으므로 3가지 경우로 구성
            zigzag_path = 0
            if direction == 'L': # 1. 왼쪽으로 들어온 경우, 오른쪽 및 시작점 탐색
                zigzag_path = max(dfs(root.right, 'R', depth + 1), dfs(root.left, 'L', 0))
            elif direction == 'R': # 2. 오른쪽으로 들어온 경우, 왼쪽 및 시작점 탐색
                zigzag_path = max(dfs(root.left, 'L', depth + 1), dfs(root.right, 'R', 0))
            else: # 3. 시작점인 경우, 왼쪽 및 오른쪽 탐색
                zigzag_path = max(dfs(root.left, 'L', depth + 1), dfs(root.right, 'R', depth + 1))
            
            return zigzag_path
        
        return dfs(root, '', -1)

            